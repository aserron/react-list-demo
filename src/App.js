import React from 'react';
import PropTypes from 'prop-types';
import withStyles from "@material-ui/core/styles/withStyles";
import {Button, Container, Grid, Paper} from "@material-ui/core";

import UserTable from "./components/user/UserTable";
import HeaderAppBar from "./components/header/HeaderAppBar";
import AddUserForm from "./components/form/AddUserForm";
import ConfirmDialog from "./components/modal/ConfirmDialog";
import AddUserDialog from "./components/modal/AddUserDialog";

/*
 * Update State
 * // Correcto
 *
 this.setState((state, props) => ({
 counter: state.counter + props.increment
 }));
 */

const styles = theme => ({
    
    
    tempFormCt: {
        background  : 'red solid',
        alignContent: "stretch"
    }
});

class App extends React.Component {
    
    constructor(props) {
        
        super(props);
        
        this.state = {
            okOpen  : false,
            formOpen: false,
            users   : [],
        }
        
        
        
        
        // // bind handlers scope
        
        // formik
        this.handleFormikSubmit = this.handleFormikSubmit.bind(this);
        this.handleOnChange = this.handleOnChange.bind(this);
        
        // confirm dialog
        this.handleOkOpen = this.handleOkOpen.bind(this);
        this.handleOkClose = this.handleOkClose.bind(this);
        
        // form dialog
        this.handleFormOpen = this.handleFormOpen.bind(this);
        this.handleFormClose = this.handleFormClose.bind(this);
        
        // debug
        this.testAdd = this.testAdd.bind(this);
        // this.handleFormSubmit   = this.handleFormSubmit.bind(this);
        
        console.info('[App.js] constructor', props, this);
    }
    
    // life cyle methods
    
    // willMount, etc
    componentDidMount() {
        
        const list = Array.from(this.props.users || []);
        
        this.setState({users: list});
        
        // console.clear();
        console.info('[App.js] componentDidMount loading list:', (list));
        
    }
    
    // eo lifecyle


// event handlers
    
    handleOnChange(e) {
        console.info("[AppNews.js].handleOnChange", e);
    }
    
  
    
    handleClearGrid() {
        // Correcto
        this.setState((state, props) => ({
            users: []
        }));
    }
    
    
    // User Methods
    
    createUser() {
        const baseUser = {
            "id"        : this.state.users.length + 1,
            "first_name": "Brand",
            "last_name" : "New",
            "email"     : "brand@columbia.edu",
            "birthday"  : "3/17/1979",
            "phone"     : "204-472-9613"
        };
        
        return Object.assign({}, baseUser);
    }
    
    addUser(user) {
        
        user = user || this.createUser();
        
        console.info('[App.js] addUser', user);
        
        let newList;
        newList = this.state.users.slice();
        newList.push(user);
        
        this.setState((state, props) => ({
            users: newList
        }));
    }
    
    
    // FORMIK METHODS
    async handleFormikSubmit(values) {
        console.log("_");
        console.info("[App.js] handleFormikSubmit ", values);
        console.log(">> submit : begin");
    
    
        // fix date
       
        
        this.setOkOpen(true);
        await new Promise((r) => setTimeout(r, 4000));
        console.log(">> submit : completed;", "\\n");
        this.setOkOpen(false);
        this.setFormOpen(false);
    
    
        // let dt = new Date();
        let dt      = values.birthday;
        let fixed = [dt.getMonth(),dt.getDate(),dt.getFullYear()].join('/');
        values.birthday = fixed;
        this.handleTestSubmit(values);
    
    }
    
    
    // OK DIALOG
    setOkOpen(isOpen) {
        this.setState({okOpen: isOpen});
    }
    
    handleOkOpen() {
        this.setState({okOpen: true});
    };
    
    handleOkClose() {
        this.setState({okOpen: false});
    };
    
    
    // Form Dialog
    
    setFormOpen(isOpen) {
        this.setState({formOpen: isOpen});
    }
    
    handleFormOpen(e) {
        console.info('[App.js] handleFormOpen:', {
            scope: this,
            event: e
        });
        this.setState({formOpen: true});
    };
    
    handleFormClose() {
        this.setState({formOpen: false});
    };
    
    // Debug Methods
    testAdd(e) {
        
        // console.clear();
        console.info('[App.js] handleTestAdd', e);
        // trying to add a user hardcoded
        
        const newUser = this.createUser();
        
        let newList;
        newList = this.state.users.slice();
        newList.push(newUser);
        
        
        // Correcto
        this.setState((state, props) => ({
            users: newList
        }));
        
        console.log("New resulting state", this.state);
    }
    
    
    handleTestSubmit(values) {
        console.warn("[App.js] handleTestSubmit", {values: values, scope: this});
        
        let user = this.createUser();
        user = Object.assign(user, values)
        this.addUser(user);
    }
    
    
    
    // Component Render
    render() {
        
        
        const DebugFooter = (props) => {
            return (
              <Container>
                  <Grid container spacing={4} style={{
                      width       : 'auto',
                      alignItems  : 'stretch',
                      alignContent: 'space-around',
                      justify     : 'space-around'
                  }}>
                      <Grid item xs={12}>
                          <Grid container spacing={2}>
                              <Grid key="1" item>
                                  <Paper className={styles.paper}>
                                      <h3>debug forced input</h3>
                                      <form onSubmit={this.handleFormikSubmit}
                                      >
                                          <input type="text" name="first_name" required/>
                                          <br/>
                                          <div className="wrapper">
                                              <Button type="submit">Submit</Button>
                                              <Button type="button" onClick={() => {
                                                  console.clear();
                                                  // console.info("Cleared Form");
                                              }}>clear[cns]
                                              </Button>
                                          </div>
                                      </form>
                                  </Paper>
                              </Grid>
                              <Grid key="2" item>
                                  <Paper className={styles.paper}>
                                      <h3>debug list</h3>
                                      <Button onClick={this.testAdd} variant='contained'>Add Test</Button>
                                      <Button onClick={this.handleClearGrid.bind(this)} variant='contained'>Clear
                                          Grid</Button>
                                      <Container>
                                          <h2>Test List</h2>
                                          <ul>
                                              {users.map((item, i, arr) => {
                                                  // console.log({id:item.id, item:item,index:i});
                                                  return (
                                                    <li key={item.id}>`{item.id} - {item.first_name} `</li>
                                                  )
                                              })}
                                          </ul>
                                      </Container>
                                  </Paper>
                              </Grid>
                              
                              <Grid key="3" item>
                                  <Paper className={styles.paper}>
                                      <h3>debug form</h3>
                                      <Container>
                                          <AddUserForm/>
                                      </Container>
                                  </Paper>
                              </Grid>
                          </Grid>
                      </Grid>
                  </Grid>
              
              </Container>
            )
        }
        
        const users = this.state.users;
        
        console.info('[App.js] render:', {
            "state.users": this.state.users,
            "props.users": this.props.users
        });
        
        
        return (
          <div className={styles.root}>
    
              <Container component="header">
                  <HeaderAppBar handleAddUser={this.handleFormOpen}/>
              </Container>
    
    
              <Container component="main">
                  <UserTable data={users}/>
              </Container>
              
              
              <Container component="footer" style={{padding: 20}}>
                  <Container><h3>Welcome user!</h3></Container>
              </Container>
              
              
              <ConfirmDialog open={this.state.okOpen}
                             handleClickOpen={this.handleOkOpen}
                             handleClose={this.handleOkClose}/>
              
              
              <AddUserDialog openModal={this.state.formOpen}
                             onCloseModal={this.handleFormClose}
              
                             handleFormikSubmit={this.handleFormikSubmit}
              
                             title          ='Add User'
                             description    =''
              />
          
          
          </div>
        );
        
    }
}

/*
 
 <
 */

App.propTypes = {
    users: PropTypes.array
};

export default withStyles(styles)(App);

