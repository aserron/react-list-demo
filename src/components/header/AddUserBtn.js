import React from "react";
import Button from "@material-ui/core/Button";
import {makeStyles} from '@material-ui/core/styles';

const useStyles = makeStyles((theme) => ({
    root: {
        flexGrow: 1,
    }
}));


const AddUserBtn = function(props) {
    
    const classes = useStyles();
    
    const defaultCb = (e) => {
        console.warn("[AddUserBtn]", "add user btn callback NOT present in props.", e)
    }
    
    // console.debug(classes);
    
    const handleAddUser = props.handleAddUser || defaultCb;
    
    return (
      <div className={classes.root}>
          <Button className     ="appbar-user-add"
                  data-testid   ="add-user-btn"
                  type="adduser"
                  variant="contained"
                  color="default"
                  onClick={handleAddUser}>
              
              Add New User
          </Button>
      </div>
    )
}

export default AddUserBtn;