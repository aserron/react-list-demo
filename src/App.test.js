import React from 'react';
import { render, screen } from '@testing-library/react';
import App from './App';
import AddUserBtn from "./components/header/AddUserBtn";


// Note: running cleanup afterEach is done automatically for you in @testing-library/react@9.0.0 or higher
// unmount and cleanup DOM after the test is finished.
// afterEach(cleanup);

// renders <App /> with a button with text.

it('renders <App /> and result isDefined.', () => {
    let result = render(<App />);
    expect(result).toBeDefined();
    
});

it('renders <App /> with a button with text.', () => {
    let result = render(<App />);
    let query  = screen.getByText('Add New User');
    
    expect(query).toBeDefined();
});

it('renders <App /> with a button with given testId', () => {
    let result = render(<App />);
    let query  = screen.getByTestId("add-user-btn");
    expect(query).toBeDefined();
});
