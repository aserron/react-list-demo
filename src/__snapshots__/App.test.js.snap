// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div>
    <header>
      <div
        class="MuiContainer-root makeStyles-root-2 MuiContainer-maxWidthLg"
      >
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
          >
            <h6
              class="MuiTypography-root makeStyles-title-4 MuiTypography-h6"
            >
              User List
            </h6>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined appbar-user-add MuiButton-colorInherit"
              tabindex="0"
              type="adduser"
            >
              <span
                class="MuiButton-label"
              >
                Add New User
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </header>
      </div>
    </header>
    <main>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <table
            aria-label="simple table"
            class="MuiTable-root makeStyles-table-5"
          >
            <thead
              class="MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  scope="col"
                >
                  ID
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                  scope="col"
                >
                  First Name
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                  scope="col"
                >
                  Last Name
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                  scope="col"
                >
                  Adult
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                  scope="col"
                >
                  Phone
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root"
            />
          </table>
        </div>
      </div>
    </main>
    <footer
      style="padding: 20px;"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4"
          style="width: auto; align-items: stretch; align-content: space-around;"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
            >
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
                >
                  <h3>
                    debug forced input
                  </h3>
                  <form>
                    <input
                      name="first_name"
                      required=""
                      type="text"
                    />
                    <br />
                    <div
                      class="wrapper"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text"
                        tabindex="0"
                        type="submit"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          Submit
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          clear[cns]
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </form>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
                >
                  <h3>
                    debug list
                  </h3>
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      Add Test
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      Clear Grid
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                  <div
                    class="MuiContainer-root MuiContainer-maxWidthLg"
                  >
                    <h2>
                      Test List
                    </h2>
                    <ul />
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
                >
                  <h3>
                    debug form
                  </h3>
                  <div
                    class="MuiContainer-root MuiContainer-maxWidthLg"
                  >
                    <form
                      action="#"
                      class="makeStyles-form-6"
                    >
                      <div
                        class="MuiFormControl-root MuiTextField-root"
                      >
                        <label
                          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
                          data-shrink="true"
                        >
                          First Name
                          <span
                            aria-hidden="true"
                            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                          >
                             *
                          </span>
                        </label>
                        <div
                          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                        >
                          <input
                            aria-invalid="false"
                            class="MuiInputBase-input MuiInput-input"
                            name="first_name"
                            placeholder="Jane"
                            required=""
                            type="text"
                            value="Mark"
                          />
                        </div>
                      </div>
                      <div
                        class="MuiFormControl-root makeStyles-formControl-7"
                      >
                        <div
                          class="MuiFormControl-root MuiTextField-root"
                        >
                          <label
                            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
                            data-shrink="true"
                          >
                            Last Name
                            <span
                              aria-hidden="true"
                              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                            >
                               *
                            </span>
                          </label>
                          <div
                            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                          >
                            <input
                              aria-invalid="false"
                              class="MuiInputBase-input MuiInput-input"
                              name="last_name"
                              placeholder="Doe"
                              required=""
                              type="text"
                              value="Konig"
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="MuiFormControl-root makeStyles-formControl-7"
                      >
                        <div
                          class="MuiFormControl-root MuiTextField-root"
                        >
                          <label
                            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
                            data-shrink="true"
                          >
                            Email
                            <span
                              aria-hidden="true"
                              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                            >
                               *
                            </span>
                          </label>
                          <div
                            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                          >
                            <input
                              aria-invalid="false"
                              class="MuiInputBase-input MuiInput-input"
                              name="email"
                              required=""
                              type="email"
                              value="andy@rotar.com"
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="MuiFormControl-root makeStyles-formControl-7"
                      >
                        <button
                          class="MuiButtonBase-root MuiButton-root MuiButton-text"
                          tabindex="0"
                          type="submit"
                        >
                          <span
                            class="MuiButton-label"
                          >
                            Submit
                          </span>
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </footer>
    <div />
  </div>
</DocumentFragment>
`;
